content/kubermatic/master/tutorials_howtos/encryption_at_rest/_index.en.md

```diff
@@ -0,0 +1,129 @@
++++
+title = "Data Encryption at Rest"
+date = 2022-05-27T12:00:00+02:00
+weight = 100
+
++++
+
+{{% notice warning %}}
+This feature is in an early stage and considered a **preview feature**. It is **not recommended** to enable this for production environments due to potential data loss.
+{{% /notice %}}
+
+To secure sensitive data stored in Kubernetes resources (e.g. `Secrets`), that data can be encrypted at rest. In Kubernetes this means that data is encrypted while stored in etcd and is only decrypted when the resource is requested via a Kubernetes API request.
+
+Data will either be encrypted with static encryption keys or via envelope encryption based on cloud provider KMS services (see [Configuring Encryption at Rest](#configuring-encryption-at-rest)). This feature is based on the [Kubernetes upstream feature for encrypting data at rest](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/).
+
+## Important Notes
+
+- Data is only encrypted _at rest_ and not when requested by users with sufficient RBAC to access it. This means that the output of `kubectl get secret <secret> -o yaml` (or similar commands/actions) remains unencrypted and is only base64-encoded. Proper RBAC management is mandatory to secure secret data at all stages.
+- Due to multiple revisions of data existing in etcd, [etcd backups]({{< ref "../etcd_backups/" >}}) might contain previous revisions of a resource that are unencrypted if the etcd backup is taken less than five minutes after data has been encrypted. Previous revisions are compacted every five minutes by `kube-apiserver`.
+
+## Configuring Encryption at Rest
+
+At the moment, encryption at rest can only be configured via KKP's Kubernetes API custom resource type `Cluster`. For example, an existing cluster can be updated via `kubectl edit cluster <Cluster ID>`. Here is an example of what `spec.encryptionConfiguration` can look like:
+
+```yaml
+# only a snippet, not valid on its own!
+spec:
+    [...]
+    encryptionConfiguration:
+        enabled: true
+        resources:
+            - secrets
+        secretbox:
+            keys:
+                - name: encryption-key-2022-01
+                  secretRef:
+                    name: encryption-key-2022-01
+                    key: key
+```
+
+Check out the [Kubermatic CRD documentation]({{< ref "../../references/crds/#encryptionconfiguration" >}}) for a full overview of the configuration options.
+
+### Encrypted Resources
+
+While most users might want to encrypt their `Secret` resources at rest, the encryption at rest feature is flexible enough to expand to other resource types. This can be useful if you have CRDs for custom resources that store sensitive data. A list of encrypted resources can be passed via `spec.encryptionConfiguration.resources`.
+
+Resource types should be passed in their plural form, all lowercase.
```

Comment:
Out of curiosity, how does resource replacement (i.e. `kubectl replace`) work when you encrypt resources that are not Secrets?

Reply:
Right now, the encryption job essentially runs `kubectl replace secrets,configmaps,machinedeployments` if you pass multiple resources. I want to make that a native Go binary later on, but the type format should stay the same, because as far as I understand that is what the EncryptionConfiguration file needs as well.